### ISSUE TITLE
Please check if my attempt to train is correct

### ISSUE THREAD

**Issue by igygi** [achalddave/segment-any-moving_detectron#1]
Hi, 

I am trying to train this model on a custom dataset. I hope you could confirm if I followed the correct steps:

Step 1: I separately trained an objectness model and a motion model, using e2e_mask_rcnn_R-50-FPN_1x.yaml as the config file for both. I also used the pretrained appearance and motion models you provided as my starting point.

Step 2: Run create_multi_input_model.py to concatenate the appearance and motion models, using e2e_mask_rcnn_R-50-FPN_1x_two_stream_adaptor.yaml as the config file for the --config argument (I'm not sure if this is correct). I saved the output of create_multi_input_model.py to merged.pth

Step 3: Run the joint model. I passed the path to the rgb and opticalflow images for --dataset, and used e2e_mask_rcc_R-50-FPN_1x_two_stream_adaptor.yaml as my config file (again, I am not sure if this is correct). 

Thank you.

**Comment by achalddave** [comment]
Hi,

It's been a while since I've trained these models, and unfortunately I have never gotten around to writing training docs. Thanks for taking a shot at training things. I believe what you've done is correct. I've gone through and pasted the commands I ran for the final model (based on log files). 

# Train objectness model
```
python tools/train_net_step.py \
--dataset coco2017objectness \
--run_name_prefix {NAME} \
--bs 8 --iter_size 2 \
--cfg configs/baselines/e2e_mask_rcnn_R-50-FPN_1x.yaml \
--use_tfboard
```

# Motion model
## Train on FT3D GT flow
```
python tools/train_net_step.py \
--dataset flyingthings3d_gtflow_train \
--run_name_prefix {NAME} \
--bs 8 \
--lr 0.002 \
--iter_size 2 \
--cfg configs/baselines/e2e_mask_rcnn_R-50-FPN_1x.yaml \
--use_tfboard
```

## Fine-tune on FT3D flownet2 flow
```
python tools/train_net_step.py \
--dataset flyingthings3d_flownet2_train \
--run_name_prefix ft3d-ft-flownet2-from-gt-train_lr-0.002 \
--bs 8 \
--lr 0.002 \
--iter_size 2 \
--cfg configs/baselines/e2e_mask_rcnn_R-50-FPN_1x.yaml \
--load_ckpt {GT_FLOW}/ckpt/model_step9999.pth \
--use_tfboard
```

# Two-stream training
## Create two-stream model architecture
```
python tools/create_multi_input_model.py \
--body-checkpoints {OBJECTNESS}/ckpt/model_step89999.pth {MOTION}/ckpt/model_step9999.pth \
--config configs/baselines/e2e_mask_rcnn_R-50-FPN_1x_two_stream_adaptor.yaml \
--head-weights-index 0 \
--num-classes 2 \
--bodymuxer-concatenateconv-init random \
--output-model {TWOSTREAM_INIT}
```

## Train joint model on YTVOS
```
python tools/train_net_step.py \
--datasets ytvos_rgb_all_moving_strict_sub_train_8-21-18 ytvos_flownet2_all_moving_strict_sub_train_8-21-18 \
--run_name_prefix {NAME} \
--bs 8 --iter_size 2 \
--cfg configs/baselines/e2e_mask_rcnn_R-50-FPN_1x_two_stream_adaptor_mask_rgb_only.yaml \
--load_ckpt {TWOSTREAM_INIT} \
--set SOLVER.BASE_LR 0.004 \
TRAIN.SNAPSHOT_ITERS 4000 \
TRAIN.FREEZE_SUBMODULES "['Conv_Body.bodies', 'Mask_Head', 'Mask_Outs']" \
SOLVER.MAX_ITER 30000 \
--use_tfboard
```

## Fine-tune on DAVIS
```
python tools/train_net_step.py \
--datasets davis16_rgb_train_instance davis16_flownet2_train_instance \
--run_name_prefix {NAME} \
--bs 8 --iter_size 2 \
--cfg configs/baselines/e2e_mask_rcnn_R-50-FPN_1x_two_stream_adaptor_mask_rgb_only.yaml \
--load_ckpt <YTVOS_TRAINED_MODEL>/ckpt/model_step4999.pth \
--set SOLVER.BASE_LR 0.004 \
TRAIN.SNAPSHOT_ITERS 4000 \
SOLVER.MAX_ITER 20000 \
TRAIN.FREEZE_SUBMODULES "['Conv_Body.bodies', 'Mask_Head', 'Mask_Outs']" \
--use_tfboard
```

**Comment by igygi** [comment]
Hi @achalddave, thank you for this. No worries, your codes were quite well documented and the accompanying log files helped a lot. I just have a clarification: 

I noticed that you used the config file e2e_mask_rcnn_R-50-FPN_1x_two_stream_adaptor_mask_rgb_only.yaml instead of  when e2e_mask_rcnn_R-50-FPN_1x_two_stream_adaptor.yaml for the joint training. What's the difference between the two?

Thanks again for this wonderful work.

**Comment by achalddave** [comment]
Great question - the "_mask_rgb_only.yaml" computes masks using just the RGB features, rather than the RGB+motion merged features. This allows us to reuse the powerful mask-branch trained on COCO. Meanwhile, the RGB+motion merged features are used by the box regression+classification heads.

**Comment by igygi** [comment]
Understood. Thanks so much @achalddave !

**Comment by ChenjieWang** [comment]
> Hi,
> 
> I am trying to train this model on a custom dataset. I hope you could confirm if I followed the correct steps:
> 
> Step 1: I separately trained an objectness model and a motion model, using e2e_mask_rcnn_R-50-FPN_1x.yaml as the config file for both. I also used the pretrained appearance and motion models you provided as my starting point.
> 
> Step 2: Run create_multi_input_model.py to concatenate the appearance and motion models, using e2e_mask_rcnn_R-50-FPN_1x_two_stream_adaptor.yaml as the config file for the --config argument (I'm not sure if this is correct). I saved the output of create_multi_input_model.py to merged.pth
> 
> Step 3: Run the joint model. I passed the path to the rgb and opticalflow images for --dataset, and used e2e_mask_rcc_R-50-FPN_1x_two_stream_adaptor.yaml as my config file (again, I am not sure if this is correct).
> 
> Thank you.



> Hi,
> 
> I am trying to train this model on a custom dataset. I hope you could confirm if I followed the correct steps:
> 
> Step 1: I separately trained an objectness model and a motion model, using e2e_mask_rcnn_R-50-FPN_1x.yaml as the config file for both. I also used the pretrained appearance and motion models you provided as my starting point.
> 
> Step 2: Run create_multi_input_model.py to concatenate the appearance and motion models, using e2e_mask_rcnn_R-50-FPN_1x_two_stream_adaptor.yaml as the config file for the --config argument (I'm not sure if this is correct). I saved the output of create_multi_input_model.py to merged.pth
> 
> Step 3: Run the joint model. I passed the path to the rgb and opticalflow images for --dataset, and used e2e_mask_rcc_R-50-FPN_1x_two_stream_adaptor.yaml as my config file (again, I am not sure if this is correct).
> 
> Thank you.

@igygi Hi, Can you tell me your configuration environment? Includes python and CUDA versions. I successfully configured flownet2 under python2.7, but python2.7 cannot run segment-any-moving. And if I use python3, I cannot successfully configure flownet2. Do you have any suggestions？ 
Thank you.

**Comment by ChenjieWang** [comment]
@achalddave Hi, can you tell me the details of how to use YTVOS to construct dataset ytvos_rgb_all_moving_strict_sub_train_8-21-18 and dataset ytvos_flownet2_all_moving_strict_sub_train_8-21-18? And what is the difference between these two datasets? I also want to know about the details of the two datasets used in Fine-tune on DAVIS.
Thanks for your wonderful work.

**Comment by achalddave** [comment]
Hi,

These are subsets of YTVOS where all the moving objects are labeled - I created these subsets by looking at each YTVOS video. I will try to release the code I used to construct the JSON files, but for now, I've uploaded the JSON files of annotations for all the datasets (YTVOS, DAVIS, FlyingThings3D) here: https://drive.google.com/open?id=1IlEs-wLEBcOUKXHdJ_K16Q2lwKrEL5Io

**Comment by ChenjieWang** [comment]
Thank you very much for sharing your datasets. With your help I believe I can start training this model soon. I hope I can reproduce the results in your paper. I look forward to your code you used to construct the JSON files.
Thanks again for your wonderful work.

**Comment by ChenjieWang** [comment]
Hi, I noticed that the two datasets of ytvos_rgb_all_moving_strict_sub_train_8-21-18 and dataset ytvos_flownet2_all_moving_strict_sub_train_8-21-18 use the same JSON file-'sub-train-no-last-frame.json'. And IM_DIR of these two datasets is /ytvos/rgb/train and /ytvos/flownet2/train, respectively. What is the difference between /ytvos/rgb/train and /ytvos/flownet2/train?

**Comment by achalddave** [comment]
The annotations are the same. The `rgb` directory should contain the original RGB frames from YTVOS, while the `flownet2` contains the computed flow for the same images.

**Comment by ChenjieWang** [comment]
Do you mean that **flownet2** stores .flow results calculated by flownet2 network? Save the .flo file locally in advance to read and train, is it for faster training?

**Comment by achalddave** [comment]
Not the .flo files, but rather png files containing the angle and magnitude, as used during evaluation. You can compute these png files using the following script: https://github.com/achalddave/segment-any-moving/blob/master/release/ytvos/compute_flow.py

Let me know if you have any other issues! If possible, could you please file a new issue for further questions?

**Comment by ChenjieWang** [comment]
Thank you for your answer. I have a issue here: https://github.com/achalddave/segment-any-moving/issues/7, but this is to ask related questions about flownet2 configuration. I can file a new issue in the next question.
